client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +146,154 @@ public void start(Listener listener) {
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server hostnames to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
+                        final InetSocketAddress socketAddress = new InetSocketAddress(address.getHostString(),
+                                address.getPort());
+                        if (!socketAddress.isUnresolved()) {
+                            resolvedAddresses.add(socketAddress);
+                        }
+                    });
+                    serverInfoGroup = ResolvedServerInfoGroup.builder().addAll(resolvedAddresses.stream()
+                            .map(ResolvedServerInfo::new)
+                            .collect(Collectors.toList()))
+                            .build();
+                }
 
                 // Update gRPC load balancer with the new set of server addresses.
                 log.info("Updating client with controllers: {}", serverInfoGroup);
                 this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
-            } catch (StatusRuntimeException e) {
-                log.warn("Failed to fetch controller addresses - {}", e);
-                this.resolverUpdater.onError(e.getStatus());
+
+                // We have found at least one controller endpoint. Repeat discovery after the regular schedule.
+                nextScheduleTimeMS = REFRESH_INTERVAL_MS;
+            } catch (Throwable e) {
+                // Catching all exceptions here since this method should never exit without rescheduling the discovery.
+                if (e instanceof StatusRuntimeException) {
+                    this.resolverUpdater.onError(((StatusRuntimeException) e).getStatus());
+                } else {
+                    this.resolverUpdater.onError(Status.UNKNOWN);
+                }
+                log.warn("Failed to construct controller endpoint list: ", e);
+
+                // Attempt retry with a lower timeout on failures to improve re-connectivity time.
+                nextScheduleTimeMS = FAILURE_RETRY_TIMEOUT_MS;
+            } finally {
+                // We avoid all blocking calls under a lock.
+                synchronized (this.$lock) {
```

Comment:
Please create a private method for this so you can use the annotation normally. (It avoids false-positives from static analysis tools)